### ISSUE TITLE
`gochecksumtype`: works only inside one package

### ISSUE THREAD

**Issue by maratori** [golangci/golangci-lint#4158]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

`gochecksumtype` doesn't report missing cases if the sum type is declared in one package and the type switch is in a different package (see https://github.com/golangci/golangci-lint/pull/3671#discussion_r1370424897).

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version v1.55.0 built with go1.21.3 from de1c3919 on 2023-10-20T11:56:32Z
```

</details>

### Configuration

<details>

```console
# paste configuration file or CLI flags here
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.21.3 linux/amd64
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/root/.cache/go-build'
GOENV='/root/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='local'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.21.3'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/app/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build2772357518=/tmp/go-build -gno-record-gcc-switches'
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
# paste output here
```

</details>

### A minimal reproducible example or link to a public repository

<details>

`/decl/decl.go`

```go
package decl

//sumtype:decl
type MySumType interface {
        sealed()
}

type VariantA struct{}

func (*VariantA) sealed() {}

type VariantB struct{}

func (*VariantB) sealed() {}
```

`/usage/usage.go`

```go
package usage

import "decl"

func DoSomething(x decl.MySumType) {
        switch x.(type) { // <-- here is expected a report about missing VariantB
        case *decl.VariantA:
        }
}
```

</details>

### Validation

- [X] Yes, I've included all information above (version, config, etc.).

**Comment by ldez** [comment]
Hello,

there is something missing in your report: Does the standalone linter has the same behavior?

After it's not unexpected based on how golangci-lint analyzes things.

**Comment by maratori** [comment]
> Does the standalone linter has the same behavior?

No, the standalone linter works well if you pass all packages to it (`go-check-sumtype ./...`).

**Comment by navijation** [comment]
Analysis drivers like golangci-lint may leverage the analysis framework's [fact interface](https://pkg.go.dev/golang.org/x/tools/go/analysis/internal/facts) to avoid rerunning analysis on packages that are not either directly or indirectly modified since the last run. What this means is that an analyzer should:

1. not strongly (caching _might_ be fine for some drivers) rely on any kind of statefulness across packages, except for state exported using the facts API
2. store _all_ information that is required across packages and not available natively from Go's type system as facts

It looks like go-check-sumtype does not leverage facts at all. So it seems incompatible golangci-lint's caching mechanism unless rewritten appropriately such as in [exhaustive](https://github.com/nishanths/exhaustive/blob/master/fact.go#L16).

**Comment by Noah-Wagner** [comment]
Same issue here. I made a minimal repro repo here: https://github.com/Noah-Wagner/golangci-example

**Comment by SKalt** [comment]
https://github.com/alecthomas/go-check-sumtype/releases/tag/v0.5.0 includes support for facts, but requires go1.27.

**Comment by ldez** [comment]
It doesn't require go1.27, this is only a wrong minimum Go version set in the `go.mod`.

I created a PR to downgrade the min go version:
- https://github.com/alecthomas/go-check-sumtype/pull/24
- https://github.com/golangci/golangci-lint/pull/6769

The `go` directive defines the _minimum version of the language_ used to write the code, so this version should never be updated unless you are requiring specific language features when writing the code.

The `toolchain` directive defines the _minimum version used to compile_ a binary, `toolchain` is not transitive, so doesn't impact library consumers.

**In the case of a library, the `go` directive should never be set to a latest stable version of Go.**

In all cases, the Go version always depends on the local version used to compile, and it will be always higher or equal to the versions defined by the `go` directive or the `toolchain` directive.
